src/Data/Avro/Decode/Lazy/FromLazyAvro.hs

```diff
@@ -0,0 +1,125 @@
+{-# LANGUAGE ConstraintKinds     #-}
+{-# LANGUAGE FlexibleInstances   #-}
+{-# LANGUAGE MultiWayIf          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Data.Avro.Decode.Lazy.FromLazyAvro
+
+where
+
+import           Control.Arrow                   (first)
+import           Data.Avro.Decode.Lazy.LazyValue as T
+import qualified Data.Avro.Encode                as E
+import           Data.Avro.HasAvroSchema
+import           Data.Avro.Schema                as S
+import qualified Data.ByteString                 as B
+import           Data.ByteString.Lazy            (ByteString)
+import qualified Data.ByteString.Lazy            as BL
+import           Data.Foldable                   (toList)
+import qualified Data.HashMap.Strict             as HashMap
+import           Data.Int
+import           Data.List.NonEmpty              (NonEmpty (..))
+import qualified Data.Map                        as Map
+import           Data.Monoid                     ((<>))
+import           Data.Tagged
+import           Data.Text                       (Text)
+import qualified Data.Text                       as Text
+import qualified Data.Text.Lazy                  as TL
+import qualified Data.Vector                     as V
+import qualified Data.Vector.Unboxed             as U
+import           Data.Word
+
+-- |  'FromLazyAvro' is a clone of 'FromAvro' except that
+-- it works for lazy values ('LazyValue').
+--
+-- Decoding from 'LazyValue` directly
+-- without converting to strict `Value` and then 'FromAvro'
+-- can be very beneficial from the performance point of view.
+class HasAvroSchema a => FromLazyAvro a where
+  fromLazyAvro :: LazyValue Type -> Result a
+
+--  | Same as '(.:)' but works on `LazyValue`.
+(.~:) :: FromLazyAvro a => HashMap.HashMap Text (LazyValue Type) -> Text -> Result a
+(.~:) obj key =
+  case HashMap.lookup key obj of
+    Nothing -> fail $ "Requested field not available: " <> show key
+    Just v  -> fromLazyAvro v
+
+instance (FromLazyAvro a, FromLazyAvro b) => FromLazyAvro (Either a b) where
+  fromLazyAvro e@(T.Union _ branch x)
+    | S.matches branch schemaA = Left  <$> fromLazyAvro x
+    | S.matches branch schemaB = Right <$> fromLazyAvro x
+    | otherwise              = badValue e "Either"
+    where Tagged schemaA = schema :: Tagged a Type
+          Tagged schemaB = schema :: Tagged b Type
+  fromLazyAvro x = badValue x "Either"
+
+instance FromLazyAvro Bool where
+  fromLazyAvro (T.Boolean b) = pure b
+  fromLazyAvro v             = badValue v "Bool"
+
+instance FromLazyAvro B.ByteString where
+  fromLazyAvro (T.Bytes b) = pure b
+  fromLazyAvro v           = badValue v "ByteString"
+
+instance FromLazyAvro BL.ByteString where
+  fromLazyAvro (T.Bytes b) = pure (BL.fromStrict b)
+  fromLazyAvro v           = badValue v "Lazy ByteString"
+
+instance FromLazyAvro Int where
+  fromLazyAvro (T.Int i) | (fromIntegral i :: Integer) < fromIntegral (maxBound :: Int)
```

Comment:
What's with the cast to `Integer`?

Reply:
Same as `FromAvro` does, it is a copy. I haven't looked at it carefully (if at all), but I _presume_ that it may have something to do with the fact that `Int` is platform-specific.
_Maybe_ it is possible that the avro value that is encoded against the `Int` schema is actually bigger than the platform-specific `Int` at the decoding time?

But I don't know the answer.

Reply:
Maybe consider using `Int32` and `Int64` specifically? `Integer` is very inefficient, so it'd be preferable to avoid going through it unnecessarily.